NEW Add ability for the upgrader to self-update

maxime-rainville opened the pull request:
This adds the ability for the upgrader to self-update. Key changes include:
* Ability to define the version number on build in a .env file that gets embedded in the PHAR executable
* Add a new self-update commands. 
   * This only works for PHAR executables. You get warnings if you try to call it on the composer install.
   * Also supports rollbacks.
* Switch to building the PHAR on PHP7.1. The idea being that it's less likely we'll install a library that relies on PHP7.2 features.
* Update deploy logic to keep an archive of older release and to compute md5 hashes.
* Added a one time warning if you are using an outdated version. This save a file in a tmp folder. So you don't keep getting the message over and over.

# Additional notes
I explored the possibility to sign our phar executable, but that would require us to publish a public key and then have the key installed along with the PHAR executable. To get this working we would need to produce an installation script like the one composer provide.

There's not any updated unit test. This is because a lot of the logic is dependent on the file being executed through a PHAR and on calling some web base service to get the latest release. So there's not really nice elegant way of unit testing this.

# Testing the updated build script
There's a condition on the `.travis.yml` file so the deploy logic only kick in if you are doing a release on the `silverstripe/silverstripe-upgrader` repo. You'll need to comment this bit out to test the build logic on a personal repo or on open-sausages.

# Parent issue
* #130

bergice commented on .env-sample:
```diff
@@ -0,0 +1 @@
+UPGRADER_VERSION=1.3.0
```
Might be better to rename `.env-sample` to `.env.example` as thats what we use in other repos.

maxime-rainville replied:
done.

bergice commented on .travis.yml:
```diff
@@ -43,9 +43,17 @@ script:
-  - if [[ -f ./upgrade-code.phar ]]; then mv upgrade-code.phar ./release/; fi
+  - if [ -d "release" ]; then rm -rf release ; fi
+  - git clone --single-branch --branch gh-pages https://github.com/$TRAVIS_REPO_SLUG.git release
+  - echo "UPGRADER_VERSION=$TRAVIS_TAG" > .env
```
Will `.env` ever not have a newline at the end of the file?

maxime-rainville replied:
The `.env` file will be built into the PHAR file. Once the PHAR file is published, we don't keep it around. So that .env file will be regenerated from scratch every time.

bergice commented on .travis.yml:
```diff
@@ -43,9 +43,17 @@ script:
+  - cp release/upgrade-code.phar release/$TRAVIS_TAG/upgrade-code.phar
+  - cd release
+  - md5sum upgrade-code.phar */upgrade-code.phar > MD5SUM
+  - cd ..
```
You can save 2 lines here by using brackets:

```
(cd release && md5sum upgrade-code.phar */upgrade-code.phar > MD5SUM)
```

maxime-rainville replied:
done.

bergice commented on bin/upgrade-code:
```diff
@@ -18,11 +18,30 @@ foreach ($paths as $path) {
+use M1\Env\Parser;
+
+// Get Upgrader version from environment file
+$version = '1.4.0';
```
Hardcoded values like this should at least be `const`. Might be ok since this is a one-off and should never change.

maxime-rainville replied:
We'll need to change it manually each time we do a release. This is for people using `composer global require silverstripe/upgrader`. Basically, I want people who run the upgrader via `composer glboal` to still get the version and to still get a warning when they need to update.

Ideally we wouldn't put that there at all. I had a look at trying to read it from the `composer.lock` file, but it sounded like too much trouble for the benefit.

I didn't make it a const because it's not something that should be available throughout the upgrader codebase.

bergice commented on doc/en/releasing.md:
```diff
@@ -1,12 +1,15 @@
+Upgrader releases are now automated with a travis build. When the tag gets built on travis, the build script will automatically run and build a PHAR executable. The PHAR executable will be attached back to the release in GitHub and publish to this repo's github pages.
+
+
+To release a new version of the upgrader follow these steps.
```
`:` instead of `.`

maxime-rainville replied:
done

bergice commented on doc/en/releasing.md:
```diff
@@ -1,12 +1,15 @@
-6. Overwrite the existing `upgrade-code.phar` with the one you built in step 3.
-7. Commit your changes to the `gh-pages` branch and push them up to GitHub. 
-8. Create a new release on GitHub and attach `upgrade-code.phar` to it.
+2. Increment the version number in `src/upgrade-code`, commit your changes and push them back up to GitHub. This will only be used for people installing the upgrader with Composer. The PHAR executable gets its version number from an environment file built into it. 
```
Did we ever get anywhere with the automated version number updating?

maxime-rainville replied:
It's automated for the PHAR build, but there's no easy way to automate it for people who install the upgrader with `composer global`.

bergice commented on doc/en/releasing.md:
```diff
@@ -1,12 +1,15 @@
-8. Create a new release on GitHub and attach `upgrade-code.phar` to it.
+2. Increment the version number in `src/upgrade-code`, commit your changes and push them back up to GitHub. This will only be used for people installing the upgrader with Composer. The PHAR executable gets its version number from an environment file built into it. 
+3. In GitHub create a new release. You should provide detail about what has changed since the last release, referencing PRs ideally.
+4. The release won't have a PHAR executable attached to it initially. You'll have to do wait for travis to complete its build. If that build fails for whatever reason, you can delete the release and unset the tag. Fix the issue and re-realease the tag.
```
Typo: `re-realease`

maxime-rainville replied:
done.

bergice commented on src/Console/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,103 @@
+            ->setDescription('Get the latest version of the SilverStripe upgrader.')
+            ->setDefinition([
+                new InputOption('yes', 'y', InputOption::VALUE_NONE, 'Do not ask for confirmation.'),
+                new InputOption('rollback', 'r', InputOption::VALUE_NONE, 'Rollback the latest self-update.')
```
Would it be worth adding specific version numbers at all? 

E.g: `upgrade-code self-update 1.4.0`

maxime-rainville replied:
I don't think the library I'm using for this support upgrading to a specific version. Just to the latest version available. In any case where trying to nudge people to always be on the latest version.

bergice commented on src/Console/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,103 @@
+        $updater = UpdateChecker::getUpdater($this->getApplication()->getVersion());
+
+        if (!$updater->hasUpdate()) {
+            $console->success('You\'re already running the latest stable release.');
```
Can we `return` here?

maxime-rainville replied:
fixed.

bergice commented on src/Console/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,103 @@
+            // This can fail for a variety of reason. We'll just let the exception bubble up.
+            $result = $updater->rollback();
+
+            if ($result) {
```
I'm seeing a lot of similarities between `update` and `rollback`. Might be worth refactoring this to keep things DRY.

maxime-rainville replied:
The update/rollback workflows ended being pretty similar, but they don't necessarily have to be.

In this context, I think it's better to keep them independent from each other if only for readability.

bergice replied:
Yep, it's hard to refactor this.

bergice commented on src/Util/UpdateChecker.php:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Get the latest upgrader version available if the message has not been displayed or false otherwise.
+     *
+     * This save the value in memory. So you don't always get the message.
```
`This saves the value in memory, so you don't always get the message.` instead of `This save the value in memory. So you don't always get the message.`

bergice replied:
Also, it is not stored in memory.

maxime-rainville replied:
`sys_get_temp_dir()` on *nix based system will return a location in `/tmp`, so it will be in memory.

On `Win` system, it returns a location in `C:\Windows\temp`. I've added a timestamp check as well because the file won't be deleted on reboot.

bergice commented on src/Console/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        if ($yes || $console->confirm("Do you want to rollback the silverstripe upgrader?", false)) {
+            // This can fail for a variety of reason. We'll just let the exception bubble up.
```
`reasons` instead of `reason`.

maxime-rainville replied:
done.

maxime-rainville commented:
@dre I think I've addressed all your points. Maybe try to get the phar build working on the open sausages repo just to make sure it works as expected. You'll need to hack`.travis.yml` a bit. You can do dummy releases there.

maxime-rainville commented:
Confirmed the autobuild works and generates a phar that can self-update. @bergice is on holiday, so I'll merged as-is.
